thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -872,6 +874,120 @@ namespace diskann {
     char *sector_scratch = query_scratch->sector_scratch;
     _u64 &sector_scratch_idx = query_scratch->sector_idx;
 
+    Timer io_timer, query_timer;
+    // cleared every iteration
+    std::vector<unsigned> frontier;
+    frontier.reserve(2 * beam_width);
+    std::vector<std::pair<unsigned, char *>> frontier_nhoods;
+    frontier_nhoods.reserve(2 * beam_width);
+    std::vector<AlignedRead> frontier_read_reqs;
+    frontier_read_reqs.reserve(2 * beam_width);
+    std::vector<std::pair<unsigned, std::pair<unsigned, unsigned *>>>
+        cached_nhoods;
+    cached_nhoods.reserve(2 * beam_width);
+
+    std::unordered_map<_u64, std::vector<_u64>> nodes_in_sectors_to_visit;
+    // Switch to brute force when filter rate is high
+    if (!bitset_view.empty() && bitset_view.count() > bitset_view.size() * kDiskAnnBruteForceFilterRate) {
```

Comment:
Let's don't make `cached_beam_search` this long. Can we make a new function for it?

Reply:
Done